### ISSUE TITLE
occasional crash in `makeTcpServer()`

### ISSUE THREAD

**Issue by kevinushey** [rstudio/httpuv#208]
To reproduce, in RStudio:

1. Create a new Plumber application (File -> New File -> Plumber API...),

2. Click the <kbd>[> Run API]</kbd> button in the Source pane's toolbar,

3. Repeatedly try clicking <kbd>[> Reload API]</kbd> to reload the application.

Eventually (I'd say 5% of the time) you should see a crash. This is the backtrace I see:
```
Thread 1 "rsession" received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
51	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007f1540c08801 in __GI_abort () at abort.c:79
#2  0x00007f15280dedf2 in uv_mutex_destroy (mutex=<optimized out>) at src/unix/thread.c:269
#3  0x00007f1528107667 in CondWait::~CondWait (this=0x7fffc98c5ff8, __in_chrg=<optimized out>) at thread.h:106
#4  Barrier::~Barrier (this=0x7fffc98c5ff0, __in_chrg=<optimized out>) at thread.h:106
#5  makeTcpServer(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, Rcpp::Function_Impl<Rcpp::PreserveStorage>, Rcpp::Function_Impl<Rcpp::PreserveStorage>, Rcpp::Function_Impl<Rcpp::PreserveStorage>, Rcpp::Function_Impl<Rcpp::PreserveStorage>, Rcpp::Function_Impl<Rcpp::PreserveStorage>, Rcpp::Function_Impl<Rcpp::PreserveStorage>, Rcpp::Vector<19, Rcpp::PreserveStorage>, Rcpp::Vector<19, Rcpp::PreserveStorage>) () at httpuv.cpp:265
#6  0x00007f15280eec8c in _httpuv_makeTcpServer () at /home/kevin/R/x86_64-pc-linux-gnu-library/3.5/Rcpp/include/Rcpp/InputParameter.h:74
#7  0x00007f1542b18310 in ?? () from /usr/lib/R/lib/libR.so
#8  0x00007f1542b188e6 in ?? () from /usr/lib/R/lib/libR.so
#9  0x00007f1542b506c2 in ?? () from /usr/lib/R/lib/libR.so
#10 0x00007f1542b5ec10 in Rf_eval () from /usr/lib/R/lib/libR.so
#11 0x00007f1542b607ef in ?? () from /usr/lib/R/lib/libR.so
#12 0x00007f1542b5edda in Rf_eval () from /usr/lib/R/lib/libR.so
#13 0x00007f1542b62546 in ?? () from /usr/lib/R/lib/libR.so
#14 0x00007f1542b5f04b in Rf_eval () from /usr/lib/R/lib/libR.so
#15 0x00007f1542b620a6 in ?? () from /usr/lib/R/lib/libR.so
#16 0x00007f1542b5f04b in Rf_eval () from /usr/lib/R/lib/libR.so
#17 0x00007f1542b607ef in ?? () from /usr/lib/R/lib/libR.so
#18 0x00007f1542b563b5 in ?? () from /usr/lib/R/lib/libR.so
#19 0x00007f1542b5ec10 in Rf_eval () from /usr/lib/R/lib/libR.so
#20 0x00007f1542b607ef in ?? () from /usr/lib/R/lib/libR.so
#21 0x00007f1542b563b5 in ?? () from /usr/lib/R/lib/libR.so
#22 0x00007f1542b5ec10 in Rf_eval () from /usr/lib/R/lib/libR.so
#23 0x00007f1542b607ef in ?? () from /usr/lib/R/lib/libR.so
#24 0x00007f1542b563b5 in ?? () from /usr/lib/R/lib/libR.so
#25 0x00007f1542b5ec10 in Rf_eval () from /usr/lib/R/lib/libR.so
#26 0x00007f1542b607ef in ?? () from /usr/lib/R/lib/libR.so
#27 0x00007f1542b5edda in Rf_eval () from /usr/lib/R/lib/libR.so
#28 0x00007f1542b620a6 in ?? () from /usr/lib/R/lib/libR.so
#29 0x00007f1542b5f04b in Rf_eval () from /usr/lib/R/lib/libR.so
#30 0x00007f1542b607ef in ?? () from /usr/lib/R/lib/libR.so
#31 0x00007f1542b5edda in Rf_eval () from /usr/lib/R/lib/libR.so
#32 0x00007f1542b8ecf2 in Rf_ReplIteration () from /usr/lib/R/lib/libR.so
#33 0x00007f1542b8f151 in ?? () from /usr/lib/R/lib/libR.so
#34 0x00007f1542b8f208 in run_Rmainloop () from /usr/lib/R/lib/libR.so
#35 0x00005616c7cb7db2 in rstudio::r::session::runEmbeddedR(rstudio::core::FilePath const&, rstudio::core::FilePath const&, bool, bool, SA_TYPE, rstudio::r::session::Callbacks const&, rstudio::r::session::InternalCallbacks*) ()
#36 0x00005616c7c9a365 in rstudio::r::session::run(rstudio::r::session::ROptions const&, rstudio::r::session::RCallbacks const&) ()
#37 0x00005616c75a9bc8 in main ()
```

```
> sessionInfo()
R version 3.5.3 (2019-03-11)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.2 LTS

Matrix products: default
BLAS: /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] httpuv_1.5.0

loaded via a namespace (and not attached):
[1] compiler_3.5.3 magrittr_1.5   R6_2.4.0       promises_1.0.1 later_0.8.0   
[6] Rcpp_1.0.0    
```

Any ideas?

**Comment by wch** [comment]
Nothing obvious comes to mind. It looks like the problem is in libuv, and we did update to libuv 1.23.1 for httpuv 1.5.0.

Are you able to reproduce this with the previous version of httpuv (1.4.5.1)?

**Comment by kevinushey** [comment]
I can reproduce with the older httpuv as well.

**Comment by wch** [comment]
I'm not able to reproduce this on my Mac with httpuv 1.5.0.

Digging into the stack trace:

The `abort()` is getting called here, because `pthread_mutex_destroy()` is returning an error:
https://github.com/rstudio/httpuv/blob/3372c2fd722c908f2e657e1eef38d4049455c741/src/libuv/src/unix/thread.c#L267-L270

Looking at the code, I don't see what could be going wrong. I'll make a version of httpuv that prints out the error code to help debug the problem.

**Comment by wch** [comment]
Note: @kevinushey was able to reproduce the problem with this simpler code:

```R
s <- startServer("0.0.0.0", 5000,
  list(
    call = function(req) {
      list(
        status = 200L,
        headers = list(
          'Content-Type' = 'text/html'
        ),
        body = "Hello world!"
      )
    }
  )
)

s$stop()
```

**Comment by kevinushey** [comment]
Some debugging with @wch indicates that this occurs when the signal / wait ordering between the main thread and background thread gets swapped. With log output for the thread ID:

```
[MAIN THREAD]: n>0: 698128320
[BG THREAD]:   n==0: 553453312
[BG THREAD]:   n>0: 553453312
[MAIN THREAD]: n==0: 698128320
EBUSY
```

**Comment by wch** [comment]
Some more detail about the problem:

The `Barrier` is created on the main thread and passed to the background thread. Both threads wait on the `Barrier`. Usually the main thread calls `.wait()` first, and then the background thread; at a finer-grained view of timing, the background thread's call to `.wait()` finishes and that thread is all done with the `Barrier`, then the main thread's call to `.wait()` finishes and it continues running, then the `Barrier` goes out of scope and its destructor is called, which in turn calls the destructor for the `CondWait` object, which calls `uv_mutex_destroy` on its mutex.

However, in some cases (more commonly on a VM), the background thread calls `.wait()` first, and then the main thread. In this case, the main thread calls `.wait()` and continues processing, and before the background thread continues and exits its `.wait()`, the main thread finishes up with its work and the `Barrier` is destroyed -- but because the background thread hasn't continued with its `.wait()`, it hasn't yet unlocked the mutex, and the `uv_mutex_destroy` aborts because it's trying to destroy a locked mutex.
